File: src/main/java/com/rimlang/rim/compiler/JarTool.java
package com.rimlang.rim.compiler;

import java.io.*;
import java.util.jar.Attributes;
import java.util.jar.JarEntry;
import java.util.jar.JarOutputStream;
import java.util.jar.Manifest;

public class JarTool {
    private final Manifest manifest = new Manifest();

    public void startManifest() {
        manifest.getMainAttributes().put(Attributes.Name.MANIFEST_VERSION, "1.0");
    }

    public JarOutputStream openJar(String jarFile) throws IOException {
        return new JarOutputStream(new FileOutputStream(jarFile), manifest);
    }

    public void addFile(JarOutputStream target, String rootPath, String source) throws IOException {
        String remaining;
        if (rootPath.endsWith(File.separator)) {
            remaining = source.substring(rootPath.length());
        } else {
            remaining = source.substring(rootPath.length() + 1);
        }
        String name = remaining.replace("\\","/");
        JarEntry entry = new JarEntry(name);
        entry.setTime(new File(source).lastModified());
        target.putNextEntry(entry);

        BufferedInputStream in = new BufferedInputStream(new FileInputStream(source));
        byte[] buffer = new byte[1024];
        while (true) {
            int count = in.read(buffer);
            if (count == -1) {
                break;
            }
            target.write(buffer, 0, count);
        }
        target.closeEntry();
        in.close();
    }

}

File: src/main/kotlin/com/mcpy/lang/compiler/JarTool.kt
package com.mcpy.lang.compiler

import java.io.BufferedInputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.util.jar.Attributes
import java.util.jar.JarEntry
import java.util.jar.JarOutputStream
import java.util.jar.Manifest

class JarTool {
    private val manifest = Manifest()
    fun startManifest() {
        manifest.mainAttributes[Attributes.Name.MANIFEST_VERSION] = "1.0"
    }

    fun openJar(jarFile: String): JarOutputStream {
        return JarOutputStream(FileOutputStream(jarFile), manifest)
    }

    fun addFile(target: JarOutputStream, rootPath: String, source: String) {
        var remaining = source.drop(rootPath.length)
        if (!rootPath.endsWith(File.separator)) {
            remaining = remaining.drop(1)
        }
        val name = remaining.replace("\\", "/")
        val entry = JarEntry(name)
        entry.time = File(source).lastModified()
        target.putNextEntry(entry)
        val stream = BufferedInputStream(FileInputStream(source))
        val buffer = ByteArray(1024)
        while (true) {
            val count = stream.read(buffer)
            if (count == -1) break
            target.write(buffer, 0, count)
        }
        target.closeEntry()
        stream.close()
    }
}